### ISSUE TITLE
Module['print'] and Module['printErr'] behaviour change from 1.37.38 to 1.38.16

### ISSUE THREAD

**Issue by vlazzarini** [kripken/emscripten#7464]
I updated to the latest release (1.38.16) and observed a couple of changes in 
Module['print'] and Module['printErr'] in the WASM build:

1. We used to be able to reassign these functions at any point, but now they can only be
assigned before the wasm binary library is initialised. This is a minor issue, as we found a
work-around. I noted changes in the generated .js file where printing used to be done by
calling these Module[]() directly, but now they seem to be assigned to 'out()' and 'err()'
functions at some earlier point, which are called by the js code when printing.

2. This is more problematic: 'printErr' is adding extra newlines at every occasion when
it prints. This breaks down our console printout formatting big time. I wonder if this is
an intended behaviour change, and if so, is there an alternative implementing the old
(no extra newlines) behaviour?

Thx a lot.

Victor
Csound team, http://csound.com

**Comment by kripken** [comment]
(1) is an intended change, yes (in 1.38.7 according to Changelog.md).

(2) Do you have a testcase showing that? It should be printing the same newlines as before, I don't think we've changed anything there.

**Comment by vlazzarini** [comment]
Thanks for the quick reply. We have made changes to accommodate (1), so it is OK. For (2),
all I have are the examples of changed behaviour, which have been tracked down to a change 
in the compiler. 

Where it is actually happening I am not quite sure, but I have a hunch. My impression is that
a long string may be parsed into separate calls to Module['printErr'] rather than a single one.

In our JS client code, we place a newline before the string is sent to the console, and if the
string message is broken down into smaller parts, the formatting will exhibit these newlines.

So in this case, you're right, the compiled code is not adding newlines, but the effect is
that these extra newlines appear. To give you an example, the following message sent
to stderr (via a vprintf() call), prints in 1.37.38 (just like it would do on a terminal)

```
rtevent:	   T 89.173 TT 89.173 M:  0.10000 0.10000
```

and in 1.38.16, we get

```
 rtevent:	   T 89.173 TT 89.173 M:
 0.10000
 0.10000
```

OTOH it could be that there are 3 vprintf() calls involved (I have not actually checked) with no newlines
in their format string, therefore printed in a single line. If the WASM compilation of each vprintf() is forcing a newline to be added, we would have the same effect. I am sorry I have not had time to look in the code to see if there are 3 calls, but I can do it later.

Does any of this ring a bell?

**Comment by vlazzarini** [comment]
I had the chance to look and yes: there are separate calls to vprintf(), and at the end of the line a separate call prints a newline. Maybe each vprintf() now is issuing a separate Module['printErr'] and
the newlines are placed by ou JS client code, whereas before some buffering would occur until a newline was seen (which would be discarded).  

However, it looked as if we still needed to add the newline in JS otherwise all strings are displayed in the same line. I'll check this.

**Comment by vlazzarini** [comment]
yes, if I remove the newline in my JS client, the messages are all in the same line, so it means that somewhere all new lines are eaten. My JS client function replaces console.log() and console.warn()
and all it does is add the string to a HTML text element. I suspect the newlines are eaten before they
are sent to console.log/warn because the console normally adds its own newlines.

In any case, it really seems to confirm that the old behaviour was doing some buffering, which 
disappeared now. This would not be a big issue if the newlines in the original strings were preserved,
and we could format based on them. At the moment, I have no means to know when a line should
end.

**Comment by kripken** [comment]
Interesting. Not sure what change on our side could cause that. If creating a testcase is not feasible, bisecting to find which change is responsible could be equally useful.

**Comment by vlazzarini** [comment]
I'll try that.

**Comment by vlazzarini** [comment]
Ok so this is getting really confusing. I wrote a simple test to do some printing, reproducing the way Csound prints stuff. It works correctly.

So I went back to the Csound code and started playing around with it to see if I could find the issue. I added this code:

```
  fprintf(stderr, "amp: %f", x);
  fprintf(stderr, "camp: %f", x);
  fprintf(stderr, "bamp: %f", x);
```

and I am getting three lines on the developer terminal. Whereas similar code in my test program prints all in a single line. At this point I can only think there is a compiler flag that I missed, as that is the only
difference that might exist. I will keep looking.

**Comment by vlazzarini** [comment]
Ok, just looked at the sequence of calls in the developer console. In my test, for each fprintf() I get

```
printChar @ test.js:1880
```

Csound gets this:

```
   flush | @ | libcsound-worklet.js:2196 
  | write | @ | libcsound-worklet.js:2104
  | write | @ | libcsound-worklet.js:4176
  | doWritev | @ | libcsound-worklet.js:4982
  | ___syscall146 | @ | libcsound-worklet.js:6315
  | ___stdio_write | @ | wasm-00b35aba-4174:60
  | _vfprintf | @ | wasm-00b35aba-4249:139
  | _fprintf | @ | wasm-00b35aba-4248:15\
```

I don't know how the code gets compiled to a different set of calls, but that's the difference. I suspect the final call to flush is the issue.

**Comment by vlazzarini** [comment]
@kripken  I finally managed to create an example demonstrating the issue. It seems to affect libraries (with the `-s LINKABLE=1` option. 

Here's the test program:

```
void msg(const char *format, ...);
  
int main() {
  msg("Hello ");
  msg( " World!");
  msg("\n");
  msg("Happy New Year!");
  msg("\n\n\n");
  msg("Bye...\n");
  return 0;
}
```

which will print to the terminal:\

```
$ ./test
Hello  World!
Happy New Year!


Bye...
```

The same example will swallow all newlines and print everything in the same line. If we add new lines in the JS client code to avoid concatenation, then each call to the `vprintf()` function will add a newline, therefore all the formatting goes askew.

I have sources and compiled binaries attached.
[print-issue.zip](https://github.com/kripken/emscripten/files/2572026/print-issue.zip)

**Comment by kripken** [comment]
Hmm, when I run that I get
```
error: Linking globals named 'msg': symbol multiply defined!
```
on the last line.

Which version are you on, @vlazzarini?

**Comment by vlazzarini** [comment]
1.38.16, build.sh runs correctly here

**Comment by vlazzarini** [comment]
Any luck with the issue, @kripken?

**Comment by kripken** [comment]
No, I'm not sure what else to try - it doesn't build for me on incoming as I wrote above. Perhaps we changed something since 1.38.16, although I can't think of anything. Does it still fail for you on the latest tag?

**Comment by vlazzarini** [comment]
Did you try just running the binary? I'll try the latest tag.

**Comment by vlazzarini** [comment]
Ok just updated to 1.38.19 and built the test here (no problems). The issue is still there. See the screenshot below:

<img width="1296" alt="scshot" src="https://user-images.githubusercontent.com/6033815/48636826-e5e9cf00-e9c3-11e8-89c6-34d07ea59296.png">

**Comment by kripken** [comment]
I didn't try the binary - if I can't build it and see the problem, the binary wouldn't be enough to help me.

Very odd that you see this on latest and I don't. Not sure what else to try on my side. Perhaps someone else can try this too, they might see something I missed or did wrong.

**Comment by vlazzarini** [comment]
I can only think you have the symbol `msg()` being defined somewhere in your build system. I've changed the function name to `display_message()` in the attached zip, maybe you can try it.
[ems.zip](https://github.com/kripken/emscripten/files/2590587/ems.zip)

**Comment by kripken** [comment]
Thanks, now I get it to build and I see the same behavior as in your screenshot, no newlines are printed out. 

I also see the same behavior without dynamic linking, and with all the C in a single file, so that all seems unrelated? The important part is `display_msg` in your HTML, which does not add newlines. If I replace it with
```
function display_msg(message) {
  console.error(message);
}
```
Then the error logging looks correct, one newline per call. That's the expected behavior for an override of `Module.printErr` - we send it entire lines, so it should print the line and a newline as necessary. This didn't change that I can recall.

**Comment by vlazzarini** [comment]
The problem is exactly that: without the newlines to format the printout I do not know when the lines end or start, therefore the formatting goes out of the window. Following your suggestion, I get

<img width="142" alt="scn2" src="https://user-images.githubusercontent.com/6033815/48651250-e305d300-e9f1-11e8-9ac4-b0fd13b2f310.png">

which is not what I would get on stdout (see how "Hello World" gets split). This is a problem for us trying to reproduce a console display correctly.

This changed sometime in between 1.37.38 and 1.38.16. In the earlier version, `printErr` would behave as expected. 

However, with the current version and not using '-s LINKABLE=1'', e.g. 

```
MacBook-Pro:ems victor$ emcc -s WASM=1 -s "BINARYEN_METHOD='native-wasm'" -s MODULARIZE=1 -s SINGLE_FILE=1 -s EXTRA_EXPORTED_RUNTIME_METHODS='["FS", "cwrap"]'  print.o main.bc -o test.js
```

the resulting code prints correctly:

<img width="181" alt="scn3" src="https://user-images.githubusercontent.com/6033815/48651497-1432d300-e9f3-11e8-90d6-faf3a7778d44.png">

The generated code in test.js is different (and much simpler).

**Comment by kripken** [comment]
I don't think we changed printErr, but perhaps you can bisect to find the commit that changed things, I may have forgotten something?

Regardless, this is intentional - you can look at Module.print and Module.printErr as being the same as puts - it gets a string, and prints it out with a newline. It's even more strict, as the string will not contain a newline. That is, the runtime calls Module.print and Module.printErr when a line is ready to be printed, and it calls the function without a newline. The reason for this behavior is that that is how console.log, console.error, etc. work in JS.

**Comment by kripken** [comment]
Sorry, I don't follow. The testcase didn't show a difference in behavior when linking or not, I tried both combinations, also as a single source file to be sure, and with and without LINKABLE.

The testcase's behavior was explained by how you were using Module.printErr, I believe. Do you see issues with printf when not customizing Module.printErr?

**Comment by vlazzarini** [comment]
Did you read my post? I explained it there. Instead of making a library, I used the ibject file directly to link to the main program and generate the test.js file. 

This showed the correct behaviour, rather than the wrong one.

It is not to do with the way I was using printErr. The implementation of printf is not correct. See my examples above when using the browser developer console.

The first string "Hello World" is broken in two separate console lines. That's not what the two vprintf() calls should be printing.

Again, it's all explained above and you can follow the instructions in the post to compile a version that works correctly.

**Comment by vlazzarini** [comment]
Just to facilitate this and be absolutely clear, see a new testcase package producing two tests (test.js and test2.js and the html pages). 
[ems.zip](https://github.com/kripken/emscripten/files/2591402/ems.zip)

The first (as above) screws the formatting. I am adding newlines in the JS client printing function:

<img width="854" alt="scrn5" src="https://user-images.githubusercontent.com/6033815/48655188-e1e19f80-ea0b-11e8-8ac1-caf52317891b.png">

Test2 prints correctly. 

<img width="550" alt="scrn4" src="https://user-images.githubusercontent.com/6033815/48655196-f45bd900-ea0b-11e8-9234-df66d84231b3.png">

As an aside, I added `-s LINKABLE=1` to the test2 built and it stopped working correctly. My hunch is that the issue is related to that flag.

**Comment by vlazzarini** [comment]
I did a bisect of released versions, and as luck would have it, version 1.38.16 introduced the issue. I have 1.38.15 happily formatting printf messages as expected. So there you go.

I  can work with 1.38.15 for the moment, but it would be good if this issue was fixed so that future upgrades don't produce buggy binaries.

**Comment by kripken** [comment]
Thanks for the more detailed testcase, now I see the problem. Sorry for not understanding earlier.

Bisecting it, I see that #7376 is to blame, which is indeed between the two versions you found.

I'm actually not sure how to fix this yet. I may be misunderstanding how musl sends an fflush to the system, but it's not obvious to me how to interpret the code in a way that would make your testcase as well as the one added in that PR both pass.

Side note, the reason the testcase behaves differently with LINKABLE=1 is that then we automatically include full filesystem support, which includes fflush.

**Comment by vlazzarini** [comment]
Thanks for this, it is not high priority for me now as 1.38.15 produces fine binaries, but it's good that you have recognised the issue and so when it's fixed I can keep up with updates.

**Comment by kripken** [comment]
Fix in #7551.

It will probably be in 1.38.21 when that is tagged.

**Comment by vlazzarini** [comment]
excellent, thank you ever so much. I must also thank you for the great software. The paper on WASM was a great read too.
